src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -748,3 +750,80 @@ impl RustcTargetData {
         self.target_config(kind).links_overrides.get(lib_name)
     }
 }
+
+/// Structure used to deal with Rustdoc fingerprinting
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RustDocFingerprint {
+    rustc_verbose_version: String,
+}
+
+impl RustDocFingerprint {
+    /// Returns the version contained by this `RustDocFingerprint` instance.
+    fn version(&self) -> &str {
+        self.rustc_verbose_version.as_str()
+    }
+
+    /// Given the `doc` dir, returns the path to the rustdoc fingerprint file.
+    fn path_to_fingerprint(doc_dir: &Path) -> PathBuf {
+        doc_dir.to_path_buf().join(".rustdoc_fingerprint.json")
+    }
+
+    /// Write the `RustDocFingerprint` info into the fingerprint file.
+    pub fn write(&self, doc_dir: &Path) -> std::io::Result<()> {
+        crate::util::paths::create_dir_all(doc_dir)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{:?}", e)))?;
+        let rustdoc_fingerprint_file =
+            File::create(RustDocFingerprint::path_to_fingerprint(doc_dir))?;
+        // We write the actual `Rustc` version to it so that we just need to compile it straight
+        // since there's no `doc/` folder to remove.
+        serde_json::to_writer(&rustdoc_fingerprint_file, &self)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{:?}", e)))
```

Comment:
Preferably, instead of using a file writer, I think it would be better just to serialize to a string and use `paths::write` to write it out. This offers better error handling, and doesn't need to bother with buffering (writing directly to a File doesn't really buffer, which doesn't interact well with serde).  There is an example [here](https://github.com/rust-lang/cargo/blob/c3abcfe8a75901c7c701557a728941e8fb19399e/src/cargo/core/compiler/fingerprint.rs#L1569-L1574).

Reply:
Thanks that was really helpful!!